### ISSUE TITLE
Flipped look-ahead attention mask with generate_square_subsequent_mask()

### ISSUE THREAD

**Issue by exfoxz** [pytorch/pytorch#34908]
## 🐛 Bug

This is a re-open of #29546. It seems to have been fixed with v.1.3.1 but now seems to have made a come back with 1.4.0a0+b9b9fd4. However, it doesn't seem to be a problem with 1.5.0a0+f9ad552, but considering 1.4.0 is the stable release, it is still an issue.

Basically the mask generated with Transformer.generate_sequare_subsequent_mask() is flipped and allows tokens at current timestep to attend to tokens in the future.

## To Reproduce
With PyTorch 1.4.0a0+b9b9fd4:
```
>>> import torch.nn as nn
>>> model = nn.Transformer()
>>> model.generate_square_subsequent_mask(4)
tensor([[0., 0., 0., 0.],
        [-inf, 0., 0., 0.],
        [-inf, -inf, 0., 0.],
        [-inf, -inf, -inf, 0.]])
```

## Expected behavior
The returned mask should be this instead (generated with PyTorch 1.5.0a0+f9ad552):
```
tensor([[0., -inf, -inf, -inf],
        [0., 0., -inf, -inf],
        [0., 0., 0., -inf],
        [0., 0., 0., 0.]])
```

## Environment

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration:
GPU 0: Tesla P100-SXM2-16GB
GPU 1: Tesla P100-SXM2-16GB
GPU 2: Tesla P100-SXM2-16GB
GPU 3: Tesla P100-SXM2-16GB

Nvidia driver version: 418.116.00
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] msgpack-numpy==0.4.3.2
[pip3] numpy==1.16.5
[pip3] numpydoc==0.9.1
[pip3] pytorch-lightning==0.6.1.dev0
[pip3] torch==1.4.0a0+b9b9fd4
[pip3] torchfile==0.1.0
[pip3] torchvision==0.6.0a0+bb5af1d
[conda] nomkl                     3.0                           0
[conda] pytorch-lightning         0.6.1.dev0               pypi_0    pypi
[conda] torch                     1.2.0a0+5a7e2cc          pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchvision               0.6.0a0+bb5af1d          pypi_0    pypi

## Additional context
Again, 1.5.0a0+f9ad552 doesn't seem to have this issue anymore, which means we have two different behaviors with two different recent versions of PyTorch.

**Comment by exfoxz** [comment]
I compiled my PyTorch from source, and my system has IBM processors, if that helps.

**Comment by zhangguanheng66** [comment]
Please check if the PR to fix the type-promotion bug is in your branch. The PR was sent after 1.3 release (October 10, 2019)

**Comment by jbschlosser** [comment]
Should this be closed out now?

**Comment by zhangguanheng66** [comment]
Confirmed. This has been fixed.
